core/src/main/scala/cats/data/Catenable.scala

```diff
@@ -0,0 +1,473 @@
+package cats
+package data
+
+import cats.implicits._
+import Catenable._
+
+import scala.annotation.tailrec
+import scala.collection.mutable.ArrayStack
+import scala.collection.immutable.SortedMap
+
+/**
+ * Trivial catenable sequence. Supports O(1) append, and (amortized)
+ * O(1) `uncons`, such that walking the sequence via N successive `uncons`
+ * steps takes O(N). Like a difference list, conversion to a `Seq[A]`
+ * takes linear time, regardless of how the sequence is built up.
+ */
+sealed abstract class Catenable[+A] {
+
+  /** Returns the head and tail of this catenable if non empty, none otherwise. Amortized O(1). */
+  final def uncons: Option[(A, Catenable[A])] = {
+    var c: Catenable[A] = this
+    val rights = new collection.mutable.ArrayBuffer[Catenable[A]]
+    // scalastyle:off null
+    var result: Option[(A, Catenable[A])] = null
+    while (result eq null) {
+      c match {
+        case Singleton(a) =>
+          val next =
+            if (rights.isEmpty) nil
+            else rights.reduceLeft((x, y) => Append(y, x))
+          result = Some(a -> next)
+        case Append(l, r) => c = l; rights += r
+        case Wrap(seq) =>
+          val tail = seq.tail
+          val next = fromSeq(tail)
+          result = Some(seq.head -> next)
```

Comment:
Can we use `(seq.head, next)` here to be clear we aren’t using any allocation for syntax enrichment.